### ISSUE TITLE
Challenge File Created Inside Application Folder

### ISSUE THREAD

**Issue by ozoak** [webprofusion/certify#135]
On running the cert request/renewal the entire folder path that would normally appear at, say, C:\iinetpub\ is being created inside a folder called "%SystemDrive%" inside the Certify application folder.

ie:
Where I expect the challenge files/etc to be created in:
```
C:\iinetpub\wwwroot\.well-known\
```
It is instead being created inside:
```
C:\Applications\Certify\%SystemDrive%\iinetpub\wwwroot\.well-known\
```

This caused the autorenew to trigger rate limiting on the cert renewal, which is far from ideal.
I've checked permissions on the c:\iinetpub\ folder path and they allow Certify (or users running it) to make changes, and Certify has been run at highest elevation privileges.

**Comment by webprofusion-chrisc** [comment]
Are other sites working or is it just that no cert request works at all? The website folder path is loaded from the IIS settings when you create the managed site in Certify. You can see the path it's trying to use in C:\ProgramData\Certify\managedsites.json - if this path is corrupted then that would suggest an issue with the reading the path info from IIS.

If you can manually renew the certificate but the auto renewal fails that suggests a different problem.

**Comment by ozoak** [comment]
IIS sites are working OK, certainly everything is functional and aside from the SSL not renewing (and the cert warnings that ensued) I can't see anything else wrong.

This is IIS though, so I am certainly not ruling anything Microsoft related out.

Looking at the managedsites.json the websiterootpath is in as:
``` "WebsiteRootPath": "%SystemDrive%\\inetpub\\wwwroot",```

**Comment by webprofusion-chrisc** [comment]
So if you request a cert manually it works? I'm trying to figure out if it's auto renew that's faulting or if you can't request a cert at all.

**Comment by ozoak** [comment]
Yeah it's all working, simply the root path in the json was wrong (or if not wrong, per se, the env var isn't being expanded, leading the bin to simply drop a folder called "%systemdrive%" into its own working path).

Once the json had been edited to read"C:\\iinetpub\\wwwroot" instead of "%SystemDrive%\\iinetpub\\wwwroot" it all worked in under 30 seconds.

**Comment by webprofusion-chrisc** [comment]
Hmm, interesting. We definitely were expanding environment variables OK previously. Possibly a regression, I'll check it out.

**Comment by DavidDeSloovere** [comment]
WebsiteRootPath should be dynamic. When deploying with tools like Octopus Deploy, the folder changes on every deploy. Certify still used the old folder, so renewal doesn't work.

**Comment by webprofusion-chrisc** [comment]
@DavidDeSloovere in the case of Octopus Deploy, would it also be recreating the IIS site? If so the Site ID within IIS would also change and at that point we would only have the domain associated with the binding to go on to match the site (which is a tenuous link).

As an alternative I believe Octopus has integrated Lets Encrypt support which may be more suitable: https://octopus.com/docs/administration/lets-encrypt-integration

**Comment by DavidDeSloovere** [comment]
@webprofusion-chrisc thanks for the fast reply.

OD can use an existing site (not changing site or app pool settings) or create/update the site (and app pool) settings to a given config.
So the Site ID will remain the same once it has been created (by OD or other).

Also note that web.config and configcheck etc need to be recreated, because the default deployment model creates a new folder (per version) with only the contents of the project (nuget package) in it.

The OD integration is only for the portal, not the sites that get created at deployment. I'm currently using that integration. The OD portal runs as selfhosted, not in IIS.
Maybe that'll come in the future. Not sure.

**Comment by webprofusion-chrisc** [comment]
The issue of website root changing over time has been previously noted for the todo list, but the idea of having it be discovered at request/renewal time is new. You would also ideally need to automate invoking the renewal process (which the certify.exe cli could do) once the site has finished deployment. The other option would be to recreate SSL binding with the current certificate rather than requesting a new one, as you may do several deployments a day in some cases.

**Comment by DavidDeSloovere** [comment]
I'm getting a sense of the complexity. Maybe this shouldn't fall onto Certifies reponsibilities then. Although it hasn't been confirmed that OD will support LE in the future, my guess is that they will get to it one day. I also just found a community step template for OD for integrating LE via ACME. Will give that a try. Thanks.

**Comment by webprofusion-chrisc** [comment]
@DavidDeSloovere no worries, you also have ACMESharp powershell scripting and letsencrypt-win-simple to choose from for scripting options.

**Comment by Marcus-L** [comment]
@DavidDeSloovere - with the next release of Certify the TLS-SNI challenge support (issue #158) may be a good alternative for you since it doesn't use the IIS website folder. Or as a current workaround the Powershell hooks are available in the current release (v2.0.13) which allows you to [reconfigure the website root dynamically](https://github.com/webprofusion/certify/blob/master/docs/Request%20Script%20Hooks.md#example-update-iis-site-root-directory-dynamically). I'll close this out, if you have any updates please open another issue.
